sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -185,6 +185,9 @@ abstract class Optimizer(catalogManager: CatalogManager)
       RemoveLiteralFromGroupExpressions,
       RemoveRepetitionFromGroupExpressions) :: Nil ++
     operatorOptimizationBatch) :+
+    // This batch rewrites data source plans and should be run after the operator
+    // optimization batch and before any batches that depend on stats.
+    Batch("Data Source Rewrite Rules", Once, dataSourceRewriteRules: _*) :+
```

Comment:
Basically, what you want to do is to add an extension point/batch between heuristics-based optimizer and cost-based optimizer. 

The batch name and comments look not good to me. We need a better name here.

Reply:
cc @maryannxue @hvanhovell

Reply:
Could you propose a name then, @gatorsmile ?

Reply:
I am open to alternatives.

Reply:
We should probably combine this batch with the one below: `earlyScanPushDownRules`, and give it a more general name similar to `extendedOperatorOptimizationRules`.

Reply:
Ping, @aokolnychyi and @gatorsmile , for the above comment.

Reply:
Thanks for getting back, @maryannxue! I think that was the original suggestion we discussed [here](https://github.com/apache/spark/pull/29066#discussion_r530663744).

However, `earlyScanPushDownRules` is exposed in a couple of places including the session builder. That's the reason I went with a new batch. I'll be happy to reconsider if we want to change all existing places.

It seems like we all agree that a better name is needed. Could you suggest one, @maryannxue?

Reply:
This also needs to run before the early filter and pushdown because those rules need to transform any new DSv2Relation into DSv2ScanRelation.

Reply:
So `dataSourceRewriteRules` looks like definitely not the appropriate name that we should expose out as an API. This is merged to Spark 3.1. Shall we revert in Spark 3.1 and think about better naming since we all agree with having a better name? specially considering "Alternatives to Breaking an API" in https://spark.apache.org/versioning-policy.html.

Reply:
It would be better to rename rather than remove it. If no one can come up with a better name, then the existing one is fine and we should use it. If there is a better name proposed, then we can simply port the rename in the 3.1 branch. No need to remove something that is useful to people that made it in before the branch was cut.

Reply:
@rdblue, it would be great if we can stick to having a good APIs as discussed and documented in "Alternatives to Breaking an API" in https://spark.apache.org/versioning-policy.html.

To clarify, I am _not_ suggesting to revert in `master` branch. I am suggesting to removing in `branch-3.1` alone to have a better designed API out. Keeping something useful to people is the exact reason why Spark happened to have a lot of cruft. We can't just remove them out for legacy reason and what we discussed in https://spark.apache.org/versioning-policy.html, but this one is not released out yet.

Reply:
I think we could have three options:

- I personally more prefer to have any name that can address these comments above, and change it in `branch-3.1` too have a good API. I filed a blocker JIRA to rename this here SPARK-33784 in case we go this way.

- If we're good with breaking it in minor releases since this is `@Unstable` API, then it works to me with AS-IS.

- Otherwise, another option is to remove this in `branch-3.1`.

I will take an action for the JIRA for which option we take.

Reply:
Something like `preCBORules`, @rdblue ?

Reply:
I'd fix the name and cherry-pick it to 3.1 rather than revert the change completely. Moreover, I'd suggest including the feature in full, not just the new batch. PR https://github.com/apache/spark/pull/30577 was submitted before 3.1 was cut as I wanted to keep the scope of PRs small and it was merged a couple of days after 3.1 branch was created. We have many places to inject custom rules but there is no way to inject rules after operator optimization. I feel it is going to be really useful for many customers. I've seen people fork session builders just because of the lack of this functionality.

Something like `postOperatorOptimizationRules` or `preCostBasedOptimizationRules` or `preCBORules` sounds OK to me.

Reply:
also cc @viirya @dbtsai @sunchao @cloud-fan @dongjoon-hyun

Reply:
@aokolnychyi, Are you saying a bad API is justified if users want?

I prefer to rename it in too. That's why I filed a blocker JIRA SPARK-33784 to keep this. I assigned this ticket to you.

To clarify again, I am suggesting to revert it in `branch-3.1` if we're in deadlock because exposing a bad API is not a good idea as discussed and documented in "Alternatives to Breaking an API" at https://spark.apache.org/versioning-policy.html.

Reply:
@HyukjinKwon, no, I meant I'd rather work on the name together to make sure it is good enough rather than revert the change. Here is a quote from my earlier comment:

> I'd fix the name and cherry-pick it to 3.1 rather than revert the change completely.

I think we are on the same page here and I'll fix the name as part of SPARK-33784. Thanks for creating it, btw!

Currently proposed names are:

- `postOperatorOptimizationRules`
- `preCostBasedOptimizationRules`
- `preCBORules`

What does everybody think about them? Any preferences?

Reply:
To me any of them works. WDYT @maryannxue @gatorsmile and @rdblue?

Reply:
I'd vote for `preCBORules`.

Reply:
According to the current discussion, could you make a follow-up PR with `preCBORules`, @aokolnychyi ? If we have a PR, it would be easier to make a final decision.

Reply:
I don't think that `preCBORules` is a good name. This batch is for rewrites that need to happen after basic optimization simplifies expressions and then pushes filters and projections. It also needs to happen before early pushdown, which in turn needs to come before CBO. All of that is before CBO, and that name doesn't capture what this is to be used for.

A more descriptive name is "planRewriteRules" because this is for rewriting plans after initial optimization, but before other optimizer rules that need to run after that rewrite, like early pushdown, CBO, etc.

The name "postOperatorOptimizationRules" is okay, but not very descriptive.

Reply:
Let's finish this discussion in a separate PR. I'll create one now.

Reply:
I created PR https://github.com/apache/spark/pull/30808